File: gherkin/go/id_generator.go
package gherkin

import (
	"github.com/gofrs/uuid"
	"strconv"
)

type IdGenerator interface {
	newId() func() string
}

type Incrementing struct {
	next int
}

func (i *Incrementing) NewId() string {
	result := strconv.Itoa(i.next)
	i.next++
	return result
}

type UUID struct {
	next int
}

func (i UUID) NewId() string {
	return uuid.Must(uuid.NewV4()).String()
}

File: gherkin/ruby/lib/gherkin/id_generator.rb
require 'securerandom'

module Gherkin
  module IdGenerator

    class IdGenerator
      # To be implemented by sub-classes
      # def new_id
      # end
    end

    class Incrementing < IdGenerator
      def initialize
        @index = -1
      end

      def new_id
        @index += 1
        @index.to_s
      end
    end

    class UUID < IdGenerator
      def new_id
        SecureRandom.uuid
      end
    end
  end
end